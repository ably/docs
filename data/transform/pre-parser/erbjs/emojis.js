const emojiRegex =
  /<% \[(.*?)\]\.each do \|emoji\| %>\s*<label><input type="radio" name="avatar" value="<%= emoji %>" \/><span class="emoji"><%= emoji %><\/span><\/label>\s*<% end %>/g;

const emojiReplacer = (content) =>
  content.replace(emojiRegex, (_match, emojis) =>
    emojis
      .split(', ')
      .map(
        (emojiWithQuotes) => {
          const cleanedEmoji = emojiWithQuotes.replace(/^'(.*)'$/, '$1');
          return `<label><input type="radio" name="avatar" value="${cleanedEmoji}" /><span class="emoji">${cleanedEmoji}</span></label>`;
        }
      )
      .join('\n'),
  );

module.exports = {
  emojiReplacer,
};