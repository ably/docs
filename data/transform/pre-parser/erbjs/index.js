const { compose } = require('lodash/fp');
const { emojiReplacer } = require('./emojis');
const { itemReplacer } = require('./item-select');
const { languageReplacer } = require('./language-map');
const { urlforReplacer } = require('./url-for');

const replaceERB = compose(emojiReplacer, languageReplacer, itemReplacer, urlforReplacer);

module.exports = {
  replaceERB,
};
