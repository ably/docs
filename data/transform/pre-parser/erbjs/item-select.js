const itemRegex =
  /<% @items\.select \{ \|d\| d\[:section\] == '(.*?)' \}\.sort_by \{ \|d\| d\[:title\] \}\.each do \|item\| %>.* "<%= html_escape\(item\[:title\]\) %>":<%= html_escape\(item\.path\) %><% end %>/gs;

const itemReplacer = (content) => content.replace(itemRegex, '<a data-for="$1">Section</a>');

module.exports = {
  itemReplacer,
};
