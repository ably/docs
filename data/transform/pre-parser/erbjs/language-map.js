const DOCUMENTATION_LANGUAGES = [
  'javascript',
  'java',
  'android',
  'kotlin',
  'python',
  'php',
  'ruby',
  'nodejs',
  'typescript',
  'objc',
  'swift',
  'go',
  'csharp',
  'cplusplus',
  'flutter',
  'c',
  'css',
  'appcelerator',
  'phonegap',
  'curl',
  'html',
  'json',
  'sh',
  'yaml',
];

const languageRegex = /<%= Ably::DOCUMENTATION_LANGUAGES.map { \|lang_id, lang\| "@#{lang_id}@" }\.join\(', '\) %>/g;

const languageReplacer = (content) =>
  content.replace(languageRegex, DOCUMENTATION_LANGUAGES.map((lang) => `@${lang}@`).join(', '));

module.exports = {
  languageReplacer,
};
