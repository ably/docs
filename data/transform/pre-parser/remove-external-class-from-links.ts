export const removeExternalClassFromLinks: StringTransformation = (content) =>
  content.replace(/"\(external\) (.*?)"/, '"$1"');
