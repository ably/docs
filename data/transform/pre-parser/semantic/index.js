const { addMinimizeForHeadings, addMinimizedIndent } = require('./minimize');

const stripComments = content => content.replace(/<!--[\s\S]*?-->\s*/g, '');

module.exports = {
    stripComments,
    addMinimizedIndent,
    addMinimizeForHeadings
}