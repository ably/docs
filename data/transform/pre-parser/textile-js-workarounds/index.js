const { compose } = require('lodash/fp');
const { fixDuplicateQuoteLinks, fixHtmlElementsInLinks } = require('./fix-links');
const { fixInlineCode } = require('./fix-inline-code');
const { fixTextileDefinitionLists } = require('./fix-textile-definition-lists');
const { addItalicisedText } = require('./add-italicised-text');
const { fixLeadingHtmlTags } = require('./fix-leading-html-tags');
const { addBoldText } = require('./add-bold-text');
const { addHyphenListSupport } = require('./add-hyphen-list-support');

// textile-js, unlike RedCloth, cannot parse multiple new lines between list items
// each list item will instead be wrapped in its own list collection
const compressMultipleNewlinesInLists = (content) => content.replace(/^(-.*)\n{2,}(?![^-])/gm, '$1\n');

// textile-js cannot parse h[1-6]. lines that are located inside another HTML tag, with leading spaces
const manuallyReplaceHTags = (content) => content.replace(/^\s*h([1-6])\.\s+(.*)$/gm, '\n<h$1>$2</h$1>');

const textileJSCompatibility = compose(
  fixLeadingHtmlTags,
  fixTextileDefinitionLists,
  fixInlineCode,
  compressMultipleNewlinesInLists,
  manuallyReplaceHTags,
  fixDuplicateQuoteLinks,
  fixHtmlElementsInLinks,
  addHyphenListSupport,
  addItalicisedText,
  addBoldText,
);

module.exports = {
  compressMultipleNewlinesInLists,
  manuallyReplaceHTags,
  fixInlineCode,
  fixTextileDefinitionLists,
  textileJSCompatibility,
};
