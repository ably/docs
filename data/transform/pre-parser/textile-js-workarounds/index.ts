import { compose } from 'lodash/fp';
import {
  fixDuplicateQuoteLinks,
  fixHtmlElementsInLinks,
  fixLinkElementsInBrackets,
  fixPunctuationInLinks,
} from './fix-links';
import { fixInlineCode } from './fix-inline-code';
import { fixTextileDefinitionLists } from './fix-textile-definition-lists';
import { addItalicisedText } from './add-italicised-text';
import { fixLeadingHtmlTags } from './fix-leading-html-tags';
import { addBoldText } from './add-bold-text';
import { makeImplicitOrderedListExplicit } from './make-implicit-ordered-list-explicit';
import { fixImgTagsWithNewlines } from './fix-html-tags-with-newlines';

// textile-js, unlike RedCloth, cannot parse multiple new lines between list items
// each list item will instead be wrapped in its own list collection
export const compressMultipleNewlinesInLists: StringTransformation = (content) =>
  content.replace(/^(-.*)\n{2,}(?![^-])/gm, '$1\n');

// textile-js cannot parse h[1-6]. lines that are located inside another HTML tag, with leading spaces
export const manuallyReplaceHTags: StringTransformation = (content) =>
  content.replace(/^\s*h([1-6])\.\s+(.*)$/gm, '\n<h$1>$2</h$1>');

export const textileJSCompatibility = compose(
  fixLeadingHtmlTags,
  fixTextileDefinitionLists,
  fixInlineCode,
  compressMultipleNewlinesInLists,
  makeImplicitOrderedListExplicit,
  manuallyReplaceHTags,
  fixImgTagsWithNewlines,
  fixDuplicateQuoteLinks,
  fixPunctuationInLinks,
  fixHtmlElementsInLinks,
  fixLinkElementsInBrackets,
  addItalicisedText,
  addBoldText,
);
