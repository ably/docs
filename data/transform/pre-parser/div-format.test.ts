import { compose } from 'lodash/fp';
import textile from 'textile-js';
import { preParser, recursivelyProcessDivs, removeNewlinesBeforeClosingTags } from '.';
import { fullyParseTextile, processTextile, textileToHtml } from '../../../data/test-utilities/process-textile';
import { replaceERB } from './erbjs';
import {
  addLanguageSupportForBlockQuotes,
  addLanguageSupportForGithubStyleCode,
  addLanguageSupportForHeadings,
  convertBlangBlocksToTokens,
  convertJSAllToNodeAndJavaScript,
  duplicateLanguageBlocks,
  enforceWhiteSpaceLevelsBetweenLanguageElements,
} from './language';
import { removeExternalClassFromLinks } from './remove-external-class-from-links';
import { addMinimizedIndent, addMinimizeForHeadings, stripComments } from './semantic';
import { textileJSCompatibility } from './textile-js-workarounds';

const misbehavingDivFixture = `

bq(definition#subscribe-event).
  default: subscribe(String name, listener("Message":#message))
  java:    void subscribe(String name, "MessageListener":#message-listener listener)
  csharp:  void Subscribe(string name, Action<"Message":#message> handler)
  ruby:    subscribe(String name) → yields "Message":#message
  objc,swift: subscribe(name: String, callback: ("ARTMessage":#message) -> Void) -> ARTEventListener?
  flutter:  StreamSubscription<"ably.Message":#message> subscribe(name: String).listen(("ably.Message":#message) -> void)

Subscribe to messages with a given event @name@ on this channel. The caller supplies <span lang="default">a listener function</span><span lang="csharp">a lambda expression</span><span lang="java">an implementation of the "MessageListener":#message-listener interface</span><span lang="ruby">a block</span>, which is called each time one or more matching messages arrives on the channel.

<div lang="objc,swift,csharp"></div>
<div lang="default">

bq(definition#subscribe-event-array).
  default: subscribe(String[] names, listener("Message":#message))
  java:    void subscribe(String[] names, "MessageListener":#message-listener listener)
  ruby:    subscribe(String *names) → yields "Message":#message
  flutter:  StreamSubscription<"ably.Message":#message> subscribe(names: List<String>).listen(("ably.Message":#message) -> void)

Subscribe a single listener to messages on this channel for multiple event @name@ values.
</div>

`;

const processedHtml = processTextile(misbehavingDivFixture);
const cheerioFinal = fullyParseTextile(misbehavingDivFixture);
describe('Ensure that divs show up correctly in all cases', () => {
  it('Ensures that divs show up correctly', () => {
    expect(recursivelyProcessDivs(misbehavingDivFixture)).toMatchInlineSnapshot(`
      "

      bq(definition#subscribe-event).
        default: subscribe(String name, listener("Message":#message))
        java:    void subscribe(String name, "MessageListener":#message-listener listener)
        csharp:  void Subscribe(string name, Action<"Message":#message> handler)
        ruby:    subscribe(String name) → yields "Message":#message
        objc,swift: subscribe(name: String, callback: ("ARTMessage":#message) -> Void) -> ARTEventListener?
        flutter:  StreamSubscription<"ably.Message":#message> subscribe(name: String).listen(("ably.Message":#message) -> void)

      Subscribe to messages with a given event @name@ on this channel. The caller supplies <span lang="default">a listener function</span><span lang="csharp">a lambda expression</span><span lang="java">an implementation of the "MessageListener":#message-listener interface</span><span lang="ruby">a block</span>, which is called each time one or more matching messages arrives on the channel.

      <div lang="objc,swift,csharp"></div>
      <notextile><div lang="default"><blockquote class="definition" id="subscribe-event-array"><p class="definition"><span lang="default">subscribe(String[] names, listener(<a href="#message">Message</a>))</span><span lang="java">void subscribe(String[] names, <a href="#message-listener">MessageListener</a> listener)</span><span lang="ruby">subscribe(String *names) → yields <a href="#message">Message</a></span><span lang="flutter">StreamSubscription&lt;<a href="#message">ably.Message</a>&gt; subscribe(names: List&lt;String&gt;).listen((<a href="#message">ably.Message</a>) &#8594; void)</span></p></blockquote><p>Subscribe a single listener to messages on this channel for multiple event <code>name</code> values.</p></div></notextile>

      "
    `);
    const partialPreParser = compose(
      // Textile compatibility must follow all other changes
      textileJSCompatibility,
      recursivelyProcessDivs,
      // Language operations
      addLanguageSupportForHeadings,
      addLanguageSupportForBlockQuotes,
      enforceWhiteSpaceLevelsBetweenLanguageElements,
      duplicateLanguageBlocks,
      addLanguageSupportForGithubStyleCode,
      convertBlangBlocksToTokens,
      convertJSAllToNodeAndJavaScript,
      // Readability/Semantic operations
      addMinimizedIndent,
      addMinimizeForHeadings,
      stripComments,
      removeNewlinesBeforeClosingTags,
      // ERB to JS
      replaceERB,
      removeExternalClassFromLinks,
    );

    expect(partialPreParser(misbehavingDivFixture)).toMatchInlineSnapshot(`
      "

      bq(definition#subscribe-event). <span lang='default'>subscribe(String name, listener(<a href="#message">Message</a>))</span><span lang='java'>void subscribe(String name, <a href="#message-listener">MessageListener</a> listener)</span><span lang='csharp'>void Subscribe(string name, Action<<a href="#message">Message</a>> handler)</span><span lang='ruby'>subscribe(String name) → yields <a href="#message">Message</a></span><span lang='objc,swift'>subscribe(name: String, callback: (<a href="#message">ARTMessage</a>) -> Void) -> ARTEventListener?</span><span lang='flutter'>StreamSubscription<<a href="#message">ably.Message</a>> subscribe(name: String).listen((<a href="#message">ably.Message</a>) -> void)</span>


      Subscribe to messages with a given event <code>name</code> on this channel. The caller supplies <span lang="default">a listener function</span><span lang="csharp">a lambda expression</span><span lang="java">an implementation of the <a href="#message-listener">MessageListener</a> interface</span><span lang="ruby">a block</span>, which is called each time one or more matching messages arrives on the channel.

      <div lang="objc,swift,csharp"></div>
      <notextile><div lang="default"><blockquote class="definition" id="subscribe-event-array"><p class="definition"><span lang="default">subscribe(String[] names, listener(<a href="#message">Message</a>))</span><span lang="java">void subscribe(String[] names, <a href="#message-listener">MessageListener</a> listener)</span><span lang="ruby">subscribe(String *names) → yields <a href="#message">Message</a></span><span lang="flutter">StreamSubscription&lt;<a href="#message">ably.Message</a>&gt; subscribe(names: List&lt;String&gt;).listen((<a href="#message">ably.Message</a>) &#8594; void)</span></p></blockquote><p>Subscribe a single listener to messages on this channel for multiple event <code>name</code> values.</p></div></notextile>

      "
    `);
    expect(preParser(misbehavingDivFixture)).toMatchInlineSnapshot(`
      "

      bq(definition#subscribe-event). <span lang='default'>subscribe(String name, listener(<a href="#message">Message</a>))</span><span lang='java'>void subscribe(String name, <a href="#message-listener">MessageListener</a> listener)</span><span lang='csharp'>void Subscribe(string name, Action<<a href="#message">Message</a>> handler)</span><span lang='ruby'>subscribe(String name) → yields <a href="#message">Message</a></span><span lang='objc,swift'>subscribe(name: String, callback: (<a href="#message">ARTMessage</a>) -> Void) -> ARTEventListener?</span><span lang='flutter'>StreamSubscription<<a href="#message">ably.Message</a>> subscribe(name: String).listen((<a href="#message">ably.Message</a>) -> void)</span>


      Subscribe to messages with a given event <code>name</code> on this channel. The caller supplies <span lang="default">a listener function</span><span lang="csharp">a lambda expression</span><span lang="java">an implementation of the <a href="#message-listener">MessageListener</a> interface</span><span lang="ruby">a block</span>, which is called each time one or more matching messages arrives on the channel.

      <div lang="objc,swift,csharp"></div>
      <notextile><div lang="default"><blockquote class="definition" id="subscribe-event-array"><p class="definition"><span lang="default">subscribe(String[] names, listener(<a href="#message">Message</a>))</span><span lang="java">void subscribe(String[] names, <a href="#message-listener">MessageListener</a> listener)</span><span lang="ruby">subscribe(String *names) → yields <a href="#message">Message</a></span><span lang="flutter">StreamSubscription&lt;<a href="#message">ably.Message</a>&gt; subscribe(names: List&lt;String&gt;).listen((<a href="#message">ably.Message</a>) &#8594; void)</span></p></blockquote><p>Subscribe a single listener to messages on this channel for multiple event <code>name</code> values.</p></div></notextile>

      "
    `);
    expect(textile(preParser(misbehavingDivFixture))).toMatchInlineSnapshot(`
      "<blockquote class="definition" id="subscribe-event">
      <p class="definition"><span lang="default">subscribe(String name, listener(<a href="#message">Message</a>))</span><span lang="java">void subscribe(String name, <a href="#message-listener">MessageListener</a> listener)</span><span lang="csharp">void Subscribe(string name, Action&lt;<a href="#message">Message</a>&gt; handler)</span><span lang="ruby">subscribe(String name) → yields <a href="#message">Message</a></span><span lang="objc,swift">subscribe(name: String, callback: (<a href="#message">ARTMessage</a>) &#8594; Void) -&gt; ARTEventListener?</span><span lang="flutter">StreamSubscription&lt;<a href="#message">ably.Message</a>&gt; subscribe(name: String).listen((<a href="#message">ably.Message</a>) &#8594; void)</span></p>
      </blockquote>
      <p>Subscribe to messages with a given event <code>name</code> on this channel. The caller supplies <span lang="default">a listener function</span><span lang="csharp">a lambda expression</span><span lang="java">an implementation of the <a href="#message-listener">MessageListener</a> interface</span><span lang="ruby">a block</span>, which is called each time one or more matching messages arrives on the channel.</p>
      <div lang="objc,swift,csharp"></div>
      <div lang="default"><blockquote class="definition" id="subscribe-event-array"><p class="definition"><span lang="default">subscribe(String[] names, listener(<a href="#message">Message</a>))</span><span lang="java">void subscribe(String[] names, <a href="#message-listener">MessageListener</a> listener)</span><span lang="ruby">subscribe(String *names) → yields <a href="#message">Message</a></span><span lang="flutter">StreamSubscription&lt;<a href="#message">ably.Message</a>&gt; subscribe(names: List&lt;String&gt;).listen((<a href="#message">ably.Message</a>) &#8594; void)</span></p></blockquote><p>Subscribe a single listener to messages on this channel for multiple event <code>name</code> values.</p></div>"
    `);
    expect(textileToHtml(misbehavingDivFixture)).toMatchInlineSnapshot(`
      "<blockquote class="definition" id="subscribe-event">
      <p class="definition"><span lang="default">subscribe(String name, listener(<a href="#message">Message</a>))</span><span lang="java">void subscribe(String name, <a href="#message-listener">MessageListener</a> listener)</span><span lang="csharp">void Subscribe(string name, Action&lt;<a href="#message">Message</a>&gt; handler)</span><span lang="ruby">subscribe(String name) → yields <a href="#message">Message</a></span><span lang="objc,swift">subscribe(name: String, callback: (<a href="#message">ARTMessage</a>) &#8594; Void) -&gt; ARTEventListener?</span><span lang="flutter">StreamSubscription&lt;<a href="#message">ably.Message</a>&gt; subscribe(name: String).listen((<a href="#message">ably.Message</a>) &#8594; void)</span></p>
      </blockquote>
      <p>Subscribe to messages with a given event <code>name</code> on this channel. The caller supplies <span lang="default">a listener function</span><span lang="csharp">a lambda expression</span><span lang="java">an implementation of the <a href="#message-listener">MessageListener</a> interface</span><span lang="ruby">a block</span>, which is called each time one or more matching messages arrives on the channel.</p>
      <div lang="objc,swift,csharp"></div>
      <div lang="default"><blockquote class="definition" id="subscribe-event-array"><p class="definition"><span lang="default">subscribe(String[] names, listener(<a href="#message">Message</a>))</span><span lang="java">void subscribe(String[] names, <a href="#message-listener">MessageListener</a> listener)</span><span lang="ruby">subscribe(String *names) → yields <a href="#message">Message</a></span><span lang="flutter">StreamSubscription&lt;<a href="#message">ably.Message</a>&gt; subscribe(names: List&lt;String&gt;).listen((<a href="#message">ably.Message</a>) &#8594; void)</span></p></blockquote><p>Subscribe a single listener to messages on this channel for multiple event <code>name</code> values.</p></div>"
    `);
    expect(processedHtml).toMatchSnapshot();
    expect(cheerioFinal).toMatchSnapshot();
  });
});
