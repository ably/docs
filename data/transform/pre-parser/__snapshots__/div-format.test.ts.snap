// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ensure that divs show up correctly in all cases Ensures that divs show up correctly 6`] = `
[
  <blockquote
    class="definition"
    id="subscribe-event"
  >
    

    <p
      class="definition"
    >
      <span
        lang="default"
      >
        subscribe(String name, listener(
        <a
          href="#message"
        >
          Message
        </a>
        ))
      </span>
      <span
        lang="java"
      >
        void subscribe(String name, 
        <a
          href="#message-listener"
        >
          MessageListener
        </a>
         listener)
      </span>
      <span
        lang="csharp"
      >
        void Subscribe(string name, Action&lt;
        <a
          href="#message"
        >
          Message
        </a>
        &gt; handler)
      </span>
      <span
        lang="ruby"
      >
        subscribe(String name) → yields 
        <a
          href="#message"
        >
          Message
        </a>
      </span>
      <span
        lang="swift"
      >
        subscribe(name: String, callback: (
        <a
          href="#message"
        >
          ARTMessage
        </a>
        ) → Void) -&gt; ARTEventListener?
      </span>
      <span
        lang="objc"
      >
        subscribe(name: String, callback: (
        <a
          href="#message"
        >
          ARTMessage
        </a>
        ) → Void) -&gt; ARTEventListener?
      </span>
      <span
        lang="flutter"
      >
        StreamSubscription&lt;
        <a
          href="#message"
        >
          ably.Message
        </a>
        &gt; subscribe(name: String).listen((
        <a
          href="#message"
        >
          ably.Message
        </a>
        ) → void)
      </span>
    </p>
    

  </blockquote>,
  <p>
    Subscribe to messages with a given event 
    <code>
      name
    </code>
     on this channel. The caller supplies 
    <span
      lang="default"
    >
      a listener function
    </span>
    <span
      lang="csharp"
    >
      a lambda expression
    </span>
    <span
      lang="java"
    >
      an implementation of the 
      <a
        href="#message-listener"
      >
        MessageListener
      </a>
       interface
    </span>
    <span
      lang="ruby"
    >
      a block
    </span>
    , which is called each time one or more matching messages arrives on the channel.
  </p>,
  <div
    lang="csharp"
  />,
  <div
    lang="swift"
  />,
  <div
    lang="objc"
  />,
  <div
    lang="default"
  >
    <blockquote
      class="definition"
      id="subscribe-event-array"
    >
      <p
        class="definition"
      >
        <span
          lang="default"
        >
          subscribe(String[] names, listener(
          <a
            href="#message"
          >
            Message
          </a>
          ))
        </span>
        <span
          lang="java"
        >
          void subscribe(String[] names, 
          <a
            href="#message-listener"
          >
            MessageListener
          </a>
           listener)
        </span>
        <span
          lang="ruby"
        >
          subscribe(String *names) → yields 
          <a
            href="#message"
          >
            Message
          </a>
        </span>
        <span
          lang="flutter"
        >
          StreamSubscription&lt;
          <a
            href="#message"
          >
            ably.Message
          </a>
          &gt; subscribe(names: List&lt;String&gt;).listen((
          <a
            href="#message"
          >
            ably.Message
          </a>
          ) → void)
        </span>
      </p>
    </blockquote>
    <p>
      Subscribe a single listener to messages on this channel for multiple event 
      <code>
        name
      </code>
       values.
    </p>
  </div>,
]
`;

exports[`Ensure that divs show up correctly in all cases Ensures that divs show up correctly 7`] = `
[
  {
    "attribs": {
      "class": "definition",
      "id": "subscribe-event",
    },
    "data": [
      {
        "data": "
",
        "name": "text",
        "type": "text",
      },
      {
        "attribs": {
          "class": "definition",
        },
        "data": [
          {
            "attribs": {
              "lang": "default",
            },
            "data": [
              {
                "data": "subscribe(String name, listener(",
                "name": "text",
                "type": "text",
              },
              {
                "attribs": {
                  "href": "#message",
                },
                "data": [
                  {
                    "data": "Message",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "a",
                "type": "tag",
              },
              {
                "data": "))",
                "name": "text",
                "type": "text",
              },
            ],
            "name": "span",
            "type": "tag",
          },
          {
            "attribs": {
              "lang": "java",
            },
            "data": [
              {
                "data": "void subscribe(String name, ",
                "name": "text",
                "type": "text",
              },
              {
                "attribs": {
                  "href": "#message-listener",
                },
                "data": [
                  {
                    "data": "MessageListener",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "a",
                "type": "tag",
              },
              {
                "data": " listener)",
                "name": "text",
                "type": "text",
              },
            ],
            "name": "span",
            "type": "tag",
          },
          {
            "attribs": {
              "lang": "csharp",
            },
            "data": [
              {
                "data": "void Subscribe(string name, Action<",
                "name": "text",
                "type": "text",
              },
              {
                "attribs": {
                  "href": "#message",
                },
                "data": [
                  {
                    "data": "Message",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "a",
                "type": "tag",
              },
              {
                "data": "> handler)",
                "name": "text",
                "type": "text",
              },
            ],
            "name": "span",
            "type": "tag",
          },
          {
            "attribs": {
              "lang": "ruby",
            },
            "data": [
              {
                "data": "subscribe(String name) → yields ",
                "name": "text",
                "type": "text",
              },
              {
                "attribs": {
                  "href": "#message",
                },
                "data": [
                  {
                    "data": "Message",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "a",
                "type": "tag",
              },
            ],
            "name": "span",
            "type": "tag",
          },
          {
            "attribs": {
              "lang": "swift",
            },
            "data": [
              {
                "data": "subscribe(name: String, callback: (",
                "name": "text",
                "type": "text",
              },
              {
                "attribs": {
                  "href": "#message",
                },
                "data": [
                  {
                    "data": "ARTMessage",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "a",
                "type": "tag",
              },
              {
                "data": ") → Void) -> ARTEventListener?",
                "name": "text",
                "type": "text",
              },
            ],
            "name": "span",
            "type": "tag",
          },
          {
            "attribs": {
              "lang": "objc",
            },
            "data": [
              {
                "data": "subscribe(name: String, callback: (",
                "name": "text",
                "type": "text",
              },
              {
                "attribs": {
                  "href": "#message",
                },
                "data": [
                  {
                    "data": "ARTMessage",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "a",
                "type": "tag",
              },
              {
                "data": ") → Void) -> ARTEventListener?",
                "name": "text",
                "type": "text",
              },
            ],
            "name": "span",
            "type": "tag",
          },
          {
            "attribs": {
              "lang": "flutter",
            },
            "data": [
              {
                "data": "StreamSubscription<",
                "name": "text",
                "type": "text",
              },
              {
                "attribs": {
                  "href": "#message",
                },
                "data": [
                  {
                    "data": "ably.Message",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "a",
                "type": "tag",
              },
              {
                "data": "> subscribe(name: String).listen((",
                "name": "text",
                "type": "text",
              },
              {
                "attribs": {
                  "href": "#message",
                },
                "data": [
                  {
                    "data": "ably.Message",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "a",
                "type": "tag",
              },
              {
                "data": ") → void)",
                "name": "text",
                "type": "text",
              },
            ],
            "name": "span",
            "type": "tag",
          },
        ],
        "name": "p",
        "type": "tag",
      },
      {
        "data": "
",
        "name": "text",
        "type": "text",
      },
    ],
    "name": "blockquote",
    "type": "tag",
  },
  {
    "attribs": {},
    "data": [
      {
        "data": "Subscribe to messages with a given event ",
        "name": "text",
        "type": "text",
      },
      {
        "attribs": {},
        "data": [
          {
            "data": "name",
            "name": "text",
            "type": "text",
          },
        ],
        "name": "code",
        "type": "tag",
      },
      {
        "data": " on this channel. The caller supplies ",
        "name": "text",
        "type": "text",
      },
      {
        "attribs": {
          "lang": "default",
        },
        "data": [
          {
            "data": "a listener function",
            "name": "text",
            "type": "text",
          },
        ],
        "name": "span",
        "type": "tag",
      },
      {
        "attribs": {
          "lang": "csharp",
        },
        "data": [
          {
            "data": "a lambda expression",
            "name": "text",
            "type": "text",
          },
        ],
        "name": "span",
        "type": "tag",
      },
      {
        "attribs": {
          "lang": "java",
        },
        "data": [
          {
            "data": "an implementation of the ",
            "name": "text",
            "type": "text",
          },
          {
            "attribs": {
              "href": "#message-listener",
            },
            "data": [
              {
                "data": "MessageListener",
                "name": "text",
                "type": "text",
              },
            ],
            "name": "a",
            "type": "tag",
          },
          {
            "data": " interface",
            "name": "text",
            "type": "text",
          },
        ],
        "name": "span",
        "type": "tag",
      },
      {
        "attribs": {
          "lang": "ruby",
        },
        "data": [
          {
            "data": "a block",
            "name": "text",
            "type": "text",
          },
        ],
        "name": "span",
        "type": "tag",
      },
      {
        "data": ", which is called each time one or more matching messages arrives on the channel.",
        "name": "text",
        "type": "text",
      },
    ],
    "name": "p",
    "type": "tag",
  },
  {
    "attribs": {
      "lang": "csharp",
    },
    "data": [],
    "name": "div",
    "type": "tag",
  },
  {
    "attribs": {
      "lang": "swift",
    },
    "data": [],
    "name": "div",
    "type": "tag",
  },
  {
    "attribs": {
      "lang": "objc",
    },
    "data": [],
    "name": "div",
    "type": "tag",
  },
  {
    "attribs": {
      "lang": "default",
    },
    "data": [
      {
        "attribs": {
          "class": "definition",
          "id": "subscribe-event-array",
        },
        "data": [
          {
            "attribs": {
              "class": "definition",
            },
            "data": [
              {
                "attribs": {
                  "lang": "default",
                },
                "data": [
                  {
                    "data": "subscribe(String[] names, listener(",
                    "name": "text",
                    "type": "text",
                  },
                  {
                    "attribs": {
                      "href": "#message",
                    },
                    "data": [
                      {
                        "data": "Message",
                        "name": "text",
                        "type": "text",
                      },
                    ],
                    "name": "a",
                    "type": "tag",
                  },
                  {
                    "data": "))",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "span",
                "type": "tag",
              },
              {
                "attribs": {
                  "lang": "java",
                },
                "data": [
                  {
                    "data": "void subscribe(String[] names, ",
                    "name": "text",
                    "type": "text",
                  },
                  {
                    "attribs": {
                      "href": "#message-listener",
                    },
                    "data": [
                      {
                        "data": "MessageListener",
                        "name": "text",
                        "type": "text",
                      },
                    ],
                    "name": "a",
                    "type": "tag",
                  },
                  {
                    "data": " listener)",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "span",
                "type": "tag",
              },
              {
                "attribs": {
                  "lang": "ruby",
                },
                "data": [
                  {
                    "data": "subscribe(String *names) → yields ",
                    "name": "text",
                    "type": "text",
                  },
                  {
                    "attribs": {
                      "href": "#message",
                    },
                    "data": [
                      {
                        "data": "Message",
                        "name": "text",
                        "type": "text",
                      },
                    ],
                    "name": "a",
                    "type": "tag",
                  },
                ],
                "name": "span",
                "type": "tag",
              },
              {
                "attribs": {
                  "lang": "flutter",
                },
                "data": [
                  {
                    "data": "StreamSubscription<",
                    "name": "text",
                    "type": "text",
                  },
                  {
                    "attribs": {
                      "href": "#message",
                    },
                    "data": [
                      {
                        "data": "ably.Message",
                        "name": "text",
                        "type": "text",
                      },
                    ],
                    "name": "a",
                    "type": "tag",
                  },
                  {
                    "data": "> subscribe(names: List<String>).listen((",
                    "name": "text",
                    "type": "text",
                  },
                  {
                    "attribs": {
                      "href": "#message",
                    },
                    "data": [
                      {
                        "data": "ably.Message",
                        "name": "text",
                        "type": "text",
                      },
                    ],
                    "name": "a",
                    "type": "tag",
                  },
                  {
                    "data": ") → void)",
                    "name": "text",
                    "type": "text",
                  },
                ],
                "name": "span",
                "type": "tag",
              },
            ],
            "name": "p",
            "type": "tag",
          },
        ],
        "name": "blockquote",
        "type": "tag",
      },
      {
        "attribs": {},
        "data": [
          {
            "data": "Subscribe a single listener to messages on this channel for multiple event ",
            "name": "text",
            "type": "text",
          },
          {
            "attribs": {},
            "data": [
              {
                "data": "name",
                "name": "text",
                "type": "text",
              },
            ],
            "name": "code",
            "type": "tag",
          },
          {
            "data": " values.",
            "name": "text",
            "type": "text",
          },
        ],
        "name": "p",
        "type": "tag",
      },
    ],
    "name": "div",
    "type": "tag",
  },
]
`;
